### PR TITLE
Debug Sylius Mailer command

### PR DESCRIPTION
**Pull request by Zales0123**
| Q               | A
| --------------- | -----
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | 
| License         | MIT

I was playing with SyliusMailerBundle a little bit yesterday and wanted to get information about emails that are configured. As it is a project using Sylius with some customizations, my first thought was "so what, I'm gonna browse yaml files like some kind of animal?!". My second thought was "but I can use `debug:config sylius_mailer` command to dump the bundle configuration which did the trick but looked rough and ugly. 

So my final resolution was - what if we have a nice debug command for the configured emails? And I wrote it. There are obviously a lot of things to improve (I still don't know if dumping the email template brings any value 🤷‍♂️), but looks nicer 🚀 

Let me know what you think about it 🫡

Zrzut ekranu 2023-01-18 o 23 30 41

PS. Thanks @loic425 for the inspiration with debug serializer command - I definitely belive following the "it's useful for me, maybe it's useful for others" path is a good idea 🖖

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -32,7 +32,8 @@
         "symfony/framework-bundle": "^5.4 || ^6.0",
         "symfony/http-kernel": "^5.4 || ^6.0",
         "twig/twig": "^2.12 || ^3.3",
-        "webmozart/assert": "^1.9"
+        "webmozart/assert": "^1.9",
+        "symfony/translation": "^5.4 || ^6.0"
```
Comment (lchrusciel) [inline composer.json:36]: Do we have to make it a hard dependency? Can't we move it to require-dev?

File: src/Bundle/Cli/DebugMailerCommand.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+namespace Sylius\Bundle\MailerBundle\Cli;
+
+use Sylius\Bundle\MailerBundle\Cli\Dumper\DumperInterface;
+use Symfony\Component\Console\Attribute\AsCommand;
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Input\InputArgument;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+
+#[AsCommand(name: 'sylius:debug:mailer', description: 'Shows configured emails and sender data')]
```
Comment (lchrusciel) [inline src/Bundle/Cli/DebugMailerCommand.php:23]: I'm not against this change, but I'm not sure we if we should use attributes in the internal code. In Symfony Commands like `\Symfony\Bundle\FrameworkBundle\Command\ConfigDebugCommand` are not using them, while for e.g. `\Symfony\Bridge\Twig\Command\DebugCommand` is using them. We preferred to not have adnotations in our code. I'm aware of the difference, but this is precedence which I would just prefer to discuss and perhaps create ADR for that

File: src/Bundle/Cli/Dumper/EmailsListDumper.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Console\Style\SymfonyStyle;
+use Symfony\Contracts\Translation\TranslatorInterface;
+
+final class EmailsListDumper implements DumperInterface
+{
+    public function __construct(
+        private array $emails,
+        private TranslatorInterface $translator,
```
Comment (lchrusciel) [inline src/Bundle/Cli/Dumper/EmailsListDumper.php:26]: Ok, now I understand why did you add translator component to mailer. And while I found it useful, I would like to suggest that perhaps adding dependency just for the sake of debugger is not enough justification. Can we make this dependency option and just don't use this feature if the component is not present?

File: src/Bundle/Cli/Dumper/EmailsListDumper.php [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+                $code,
+                $emailConfiguration['template'],
+                $emailConfiguration['enabled'] ? 'yes' : 'no',
+                isset($emailConfiguration['subject']) ? $this->translator->trans($emailConfiguration['subject']) : '',
+            ];
+        }
+
+        $io->section('<info>Emails</info>');
+
+        $table = new Table($output);
```
Comment (loic425) [inline src/Bundle/Cli/Dumper/EmailsListDumper.php:46]: There's a Symfony style for tables. I think it's be better to use it.

File: src/Bundle/Resources/config/services.xml [hunk 5]
```diff
@@ -86,5 +86,27 @@
         >
             <argument type="tagged" tag="sylius_mailer.email_modifier" />
         </service>
+
+        <service id="Sylius\Bundle\MailerBundle\Cli\Dumper\EmailDetailsDumper">
```
Comment (loic425) [inline src/Bundle/Resources/config/services.xml:90]: Following @mbabker advices, I'm not sure if it'd be better to follow Symfony best practices for service ids instead.

File: src/Bundle/tests/Cli/DebugMailerCommandTest.php [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    /** @test */
+    public function it_lists_all_configured_emails_and_sender_data(): void
+    {
+        $this->commandTester->execute([]);
+        $this->commandTester->assertCommandIsSuccessful();
+
+        $output = $this->commandTester->getDisplay();
+        $this->assertStringContainsString('Name    Sender', $output);
```
Comment (loic425) [inline src/Bundle/tests/Cli/DebugMailerCommandTest.php:40]: Why don't you test directly the whole input instead?

File: src/Bundle/Cli/DebugMailerCommand.php [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+final class DebugMailerCommand extends Command
+{
+    public function __construct(
+        private DumperInterface $senderDataDumper,
+        private DumperInterface $emailsListDumper,
+        private DumperInterface $emailDetailDumper,
+    ) {
```
Comment (lchrusciel) [inline src/Bundle/Cli/DebugMailerCommand.php:30]: While I understand the reasoning behind single interface, I would suggest splitting them. `emailDetailDumper` is not working without `code` inside of data array, while `senderDataDumper` and `emailsListDumper` are not using the last argument at all. All together, the order of injection (to some degree) matters and therefore I would suggest two different interfaces. 

As we are discussing this part (and being really picky) In addition, lest maybe use iterator for `DumperInterface`, so once someone would like to add a new one it would be enough to add tag or just extract it to separate service with two calls inside. But TBH probably no one will use this extendibility point

File: src/Bundle/Resources/config/services.xml (hunk 5 above)
Comment (lchrusciel) [reply src/Bundle/Resources/config/services.xml:90]: Splitting interfaces and grouping one service behind composite would solve this problem :+1:

File: src/Bundle/Cli/DebugMailerCommand.php (hunk 2 above)
Comment (loic425) [reply src/Bundle/Cli/DebugMailerCommand.php:23]: I was wondering this too... Yeah, maybe an ADR could be cool. Cause Symfony use them a lot in the base code, I think that's ok

File: src/Bundle/Cli/Dumper/EmailsListDumper.php (hunk 3 above)
Comment (loic425) [reply src/Bundle/Cli/Dumper/EmailsListDumper.php:26]: I agree with Lukasz.

File: src/Bundle/Cli/DebugMailerCommand.php (hunk 2 above)
Comment (mbabker) [reply src/Bundle/Cli/DebugMailerCommand.php:23]: To support lazy loaded commands (which prevents instantiating command classes needlessly in a lot of cases), you either need to use the `AsCommand` attribute or set the right attributes on the `console.command` service tag.  If the `AsCommand` attribute is good enough for the core Symfony commands (on one of my Symfony 6.2 apps, of the roughly ~50 references to the `console.command` tag, only 3 have any extra attributes set), there's no reason they can't be good enough here too.

**Comment by Zales0123** [comment]
Closing in favor of https://github.com/Sylius/SyliusMailerBundle/pull/223 🖖
